core/save-xml.cpp

```diff
@@ -236,7 +236,10 @@ static void save_sample(struct membuffer *b, struct sample *sample, struct sampl
 {
 	int idx;
 
-	put_format(b, "  <sample time='%u:%02u min'", FRACTION_TUPLE(sample->time.seconds, 60));
+	if (sample->time.ms == 0)
+		put_format(b, "  <sample time='%u:%02u min'", FRACTION_TUPLE(sample->time.seconds, 60));
+	else
+		put_format(b, "  <sample time='%u:%02u.%03u min'", FRACTION_TUPLE(sample->time.seconds, 60), sample->time.ms);
```

Comment:
This breaks backwards compatibility with all existing versions of Subsurface. This will likely lead to a flurry of bug reports from users of dive computers reporting sub-second timestamps when they find out that they are unable to open their log files with Subsurface on some of their devices.

It also drops feature parity between file based and cloud based log files, likely leading to another flurry of bug reports.

Reply:
> This breaks backwards compatibility with all existing versions of Subsurface. This will likely lead to a flurry of bug reports from users of dive computers reporting sub-second timestamps when they find out that they are unable to open their log files with Subsurface on some of their devices.
> 
> It also drops feature parity between file based and cloud based log files, likely leading to another flurry of bug reports.

I've created my logbook with this version, synced it to the cloud, edited it with the mobile version, synced it back, and opened it with the latest stable version. I couldn't see any problem.

You're right that while storing this to the cloud, the precision is lost again. I forgot to check that and will have a look.

Reply:
After the latest commit I added we also export the sub-second part when storing to Git and we're able to read it again.

As expected the precision is still lost when storing it with an older version in between, but "it works"^TM.

Reply:
This does not work for me - after importing a dive with ms resolution timestamps, saving it to the cloud, closing the log, opening the cloud storage again, and saving the previously imported dive into a file the ms resolution is gone.

Reply:
Indeed, I missed that. Should be fixed.

Reply:
Thanks for the review and thorough testing :)

I still have to get used to the code base and all the use cases and flows that I'm not aware of yet.